### ISSUE TITLE
Please un-protect `ChangeNotifier.notifyListeners`

### ISSUE THREAD

**Issue by lukehutch** [flutter/flutter#135478]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

I often create a `ValueListener<T>`, where `T` is a class or collection whose members may change, even if the object of type `T` never changes. I want to be able to notify the listeners of these changes. For example, if I create a `ValueListener<Set<int>> valueListener`, I should be able to call:

```dart
if (valueListener.value.add(5)) {
  valueListener.notifyListeners();
}
```

However, `ValueListener`'s superclass `ChangeNotifier` marks `notifyListeners()` as `@protected`.

I can still call the `notifyListeners` method, with a linter override directive, but I don't understand why this method is not public to start with.

### Proposal

Please make `ChangeNotifier.notifyListeners` a public method.

**Comment by adil192** [comment]
+1 to this, even though it's been brought up several times already (e.g. https://github.com/flutter/flutter/issues/27448 and https://github.com/flutter/flutter/issues/29958).

I often need to use a `ValueNotifier<List>` and there's currently no good way to tell the ValueNotifier that the List has changed.

An interesting proposal that I haven't seen anybody mention yet is to make Lists and Sets and other Dart objects listenable by default, so we can do something like:
```dart
List<String> selectedFiles = [];

@override
Widget build(BuildContext build) {
  return ValueListenableBuilder(
    valueListenable: selectedFiles,
    // ...
  );
}
```

Currently, my solution is to write an extension like this and then use `notifyListenersPlease` instead of `notifierListeners` so you only need to ignore the warning in one place.
```dart
import 'package:flutter/material.dart';

extension ChangeNotifierExtensions on ChangeNotifier {
  /// This is a hack to allow us to call [notifyListeners]
  /// which is usually protected.
  /// 
  /// In my (@adil192) opinion, [notifyListeners] should be public. See
  /// - https://github.com/flutter/flutter/issues/135478
  /// - https://github.com/flutter/flutter/issues/27448
  /// -  https://github.com/flutter/flutter/issues/29958
  /// - etc.
  void notifyListenersPlease() {
    // ignore: invalid_use_of_protected_member, invalid_use_of_visible_for_testing_member
    notifyListeners();
  }
}
```

**Comment by lukepighetti** [comment]
ChangeNotifier is an awesome way to build simple triggers and having notifyListeners protected gets in the way of using them as such

**Comment by feinstein** [comment]
IMO this can lead to bugs, since all parts of your code can trigger notifications, even when they shouldn't. The only piece of code that knows when to trigger the notifications safely is the class that implements the ChangeNotifier.

It's safer IMO to keep it as is, and if someone feel it's safe to expose it, than it's easy to add a quick boilerplate for this.

That's my unpopular 2 cents.

**Comment by feinstein** [comment]
Let's say you have a ChangeNotifier that exposes 3 String getters. Each string comes from an API. This ChangeNotifier should only notify when all 3 strings have been pulled from the API. If you trigger the refresh in the middle of the API call, then your UI could show 1 new string and 2 old strings, which could be a bug.

**Comment by lukehutch** [comment]
No, it's not correct to assume that even if all 3 values have been fetched separately and asy chronously, that the local view of the three values is now consistent with the server's view of the three values, since the server values may change after one or two calls have completed.

In other words, you have not correctly identified the source of the problem. If three values are supposed to be consistent, they should be fetched in a single call.

ChangeNotifier is supposed to fire every time something (anything) has changed. One of the very reasons to expose notifyListeners is so that the ChangeNotifier can override this behavior, and choose to defer notifying listeners in exactly the situation you bring up (by waiting until all relevant changes of a set of changes have been made before notifying listeners).

**Comment by feinstein** [comment]
I agree it should be on the same call, I was just trying to come up with a simple example about how a state manager knows when things are ready to be consumed, but other classes won't know. 

If any class can trigger the notification from the state manager, they can impact all listeners, be they widgets or not, as many people use ChangeNotifier as an observable.

As far as I can see, this sounds to me as a case where we have one class that abstracts a behaviour (the class that extends the ChangeNotifier). And another class that has no clue what the other does internally, but still is letting all listeners know that there are changes to be consumed.

The structure doesn't seem right to me. Object A needs an update from Object B, but it will trigger an update on objects C, D and more by invoking a notification. I feel it's safer if only Object B has the ability to notify its listeners, as only it knows when things are ready.

You are right that modifications SHOULD be done in 1 call, but I am sure many people didn't do it, specially because they know they can be more relaxed about it, since they control the notifications. If this change is made, then all code that didn't update all variables at once is vulnerable to bugs.

**Comment by lukepighetti** [comment]
> IMO this can lead to bugs, since all parts of your code can trigger notifications, even when they shouldn't. 

“someone could write a bug” is too general a concern imo

otoh, right now you can construct a ChangeNotifier() that you can register callbacks on but cannot trigger anything. Half of the API to use ChangeNotifier is protected. if we care so much about safety we should make non-constructable

**Comment by feinstein** [comment]
I think this makes sense for ValueNotifier, as the name and api implies its just a value holder, but IMO this doesn't apply to ChangeNotifier, as it's name implies a more generic mechanism, and it can be used in many ways.

Maybe ValueNotifier should be independent of ChangeNotifier?

PS: Time, GPS, etc, can all be state.

**Comment by lukepighetti** [comment]
> I think this makes sense for ValueNotifier, as the name and api implies its just a value holder, but IMO this doesn't apply to ChangeNotifier, as it's name implies a more generic mechanism, and it can be used in many ways.
> 
> Maybe ValueNotifier should be independent of ChangeNotifier?
> 
> PS: Time, GPS, etc, can all be state.

ChangeNotifier is just a primitive reactive class. I don't see any reason why it should be hamstrung compared to it's subclasses

**Comment by feinstein** [comment]
I don't use ChangeNotifiers anymore, but I did in the past, so now it's 0, but in the past I would say 7. They were my state managers, with lots of getters, and I wouldn't mind updating fields as soon as async calls returned, and notifyjng at the end, since I knew I had all the control inside the state manager.

I am not opposing without understanding your pains. I am opposing bc I can see many bugs arising from this, specially from junior devs.

OOP sometimes can bite us, ChangeNotifier is a reactive primitive to ValueNotifier, but their meanings are so different that introducing this simple change can make sense for one but not the other, that's why I am suggesting maybe splitting them. It's like Bird > Kiwi, and Bird defines `fly()`. It makes sense for mosr birds, but not for kiwi.

How about we ask the original author why he made it protected? Because my bet is he saw this same scenario I am pointing out.

I better not comment anymore bc it looks like I am insisting, and I am just wanting to give my 2 cents.
